nexus/types/src/deployment/clickhouse.rs

```diff
@@ -0,0 +1,118 @@
+    /// polling. However, we never want to use old configuration if we have
+    /// already seen and acted on newer configuration.
+    pub highest_seen_keeper_leader_committed_log_index: u64,
+
+    /// The desired state of the clickhouse keeper cluster
+    ///
+    /// We decouple deployment of zones that should contain clickhouse keeper
+    /// processes from actually starting or stopping those processes, adding or
+    /// removing them to/from the keeper cluster, and reconfiguring other keeper
+    /// and clickhouse server nodes to reflect the new configuration.
+    ///
+    /// As part of this decoupling, we keep track of the intended zone
+    /// deployment in the blueprint, but that is not enough to track the desired
+    /// state of the keeper cluster. We are only allowed to add or remove one
+    /// keeper node at a time, and therefore we must track the desired state of
+    /// the keeper cluster which may change multiple times until the keepers in
+    /// the cluster match the deployed zones. An example may help:
+    ///
+    ///   1. We start with 3 keeper nodes in 3 deployed keeper zones and need to
+    ///      add two to reach our desired policy of 5 keepers
+    ///   2. The planner adds 2 new keeper zones to the blueprint
+    ///   3. The planner will also add **one** new keeper process that matches
+    ///      one of the deployed zones to the desired keeper cluster.
+    ///   4. The executor will start the new keeper process, attempt to add it
+    ///      to the keeper cluster by pushing configuration updates to the other
+    ///      keepers, and then updating the clickhouse server configurations to
+    ///      know about the new keeper.
+    ///   5. If the keeper is successfully added, as reflected in inventory, then
+    ///      steps 3 and 4 above will be retried for the next keeper process.
+    ///   6. If the keeper is not successfully added by the executor it will
+    ///      continue to retry indefinitely.
+    ///   7. If the zone is expunged while the planner is has it as part of its
+    ///      desired state, and the executor is trying to add it, the keeper
+    ///      will be removed from the desired state in the next blueprint. If it
+    ///      has been added by an executor in the meantime it will be removed on
+    ///      the next iteration of an executor.
+    pub keepers: BTreeMap<OmicronZoneUuid, KeeperId>,
+
+    /// The desired state of clickhouse server processes on the rack
+    ///
+    /// Clickhouse servers do not have the same limitations as keepers
+    /// and can be deployed all at once.
+    pub servers: BTreeMap<OmicronZoneUuid, ServerId>,
+}
+
+impl ClickhouseClusterConfig {
+    pub fn new(cluster_name: String) -> ClickhouseClusterConfig {
+        ClickhouseClusterConfig {
+            generation: Generation::new(),
+            max_used_server_id: 0.into(),
+            max_used_keeper_id: 0.into(),
+            cluster_name,
```

Comment:
What do you think about using this constant instead of taking cluster_name as a parameter? 

 https://github.com/oxidecomputer/omicron/blob/049663724ecf779e43dad7908a16f07055a68a4e/clickhouse-admin/types/src/lib.rs#L18

Reply:
I don't think that's necessarily going to work in the future multi-rack world, but it might!  I can envision customers having separate clusters for different racks or fleets and wanting to distinguish between them possibly. Maybe we wouldn't tie the cluster name that way, but I wouldn't guarantee keeping them the same indefinitely.

In any case, I don't think it hurts to keep it flexible for now. For this configuration we are only going to set the name once and carry it forward indefinitely. At the call site we can pass in the constant.